src/box/relay_file_join.h

```diff
@@ -0,0 +1,46 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2024, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include "trivia/config.h"
+
+#if defined(ENABLE_FILE_JOIN)
+#include "relay_file_join_impl.h"
+#else /* !defined(ENABLE_FILE_JOIN) */
+
+/**
+ * Send initial JOIN rows from files to the replica. For memtx .snap is sent,
+ * for vinyl - .run files.
+ *
+ * @param io client connection
+ * @param stream relay's stream, to which data is sent
+ * @param sync sync from incoming JOIN request
+ * @param vclock[out] vclock of the snapshot
+ * @param replica_version_id peer's version
+ * @param cursor_engine code of the cursor engine
+ * @param cursor_vclock vclock of the snapshot to fetch from
+ * @param cursor_lsn start lsn in file, engine specific
```

Comment:
AFAIU the problem you're trying to solve here is avoiding rebootstrap of a CDC client if it fails while fetching a snaphot. One way to achieve that is sending a snapshot because we can seek a position in it. There's another way though: we could leave the join iterator lingering for a while after a failure (say for 1 minute) to let the recipient reconnect and resume from where it stopped. Did you consider it? This could be done with the regular join, without introducing "file join".

Reply:
> There's another way though: we could leave the join iterator lingering for a while after a failure (say for 1 minute) to let the recipient reconnect and resume from where it stopped.

This requires another RFC, we don't have time for that already. File snapshot fetching is already described in approved: [link](https://www.notion.so/tarantool/File-snapshot-fetching-443746c172274086be63c942f7fbd36f). Moreover, file snapshot fetching strongly connected to the persistent GC feature, we're currently implementing. 

The major advantage of files is that they don't take place in RAM unlike read-views. If we start saving read views, we'll have to deal with the maximum memory, which may be consumed by read-views, in order to avoid memory overflow. Files are ok, IMHO

Reply:
Tbh, I like Vova's idea. I don't think we considered it? The fact that the RFC is approved already is true, and it will allow us to get back to the file approach fast, but I would still be interested how much simpler the "lingering" cursors would be.

Memory-wise I wouldn't worry. We can limit the max number of cursors being open if it ever becomes an issue. Right now it wasn't ever a problem and it already works this way when joining not from a file. Would a cursor die after one minute or right away won't matter if join itself takes an hour or more. If it takes less, then the dataset is small, and it is as well unlikely to cause memory size issues.

Reply:
> I don't think we considered it?

We didn't. Though, we'll be able to do it with the same API, proposed in RFC. I see no reason to write a new one, it's a question of implementation.

> Would a cursor die after one minute or right away won't matter if join itself takes an hour or more.

We'll have to allow client to configure how much time read view lives. So, if some big value is set, it may become an issue. The question, whether we should deal with this right now or later. IMHO, two new box.cfg options for read views are needed - max_memory, taken by all read views, and retention period.

---

@sergepetrenko, what do you think?  Whether I should write a new version with read views? It indeed seems it'll be easier in terms of implementation and subsequent support, as current implementation of vinyl file join is way to difficult and I see no way to simplify it. However, the current implementation with files works and fully complies with the API from RFC.

Reply:
With read views we will also be able to drop opts from `_gc_consumers` and this cringe `with_snap` option. I'm voting for spending some time now and implementing read view retention period, as I don't want to support vinyl file join)